### PR TITLE
[BEAM-4145] Populate the worker_id metadata in the Java SDK Harness

### PR DESCRIPTION
**Pull request by tgroh**
The ID is required for any control service with multiple clients, which is the general case.

Enable `ManagedChannelFactory` implementations to register interceptors on all of the
created channels.

Add an interceptor which attaches the ID to the metadata. Use this interceptor in the 
Sdk Harness Control Client.

Populate the ID passed via the container contract in the boot go code.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @angoenka @youngoli

**Comment by tgroh** [comment]
run java precommit

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 1]
```diff
@@ -41,7 +42,24 @@ public static ManagedChannelFactory createEpoll() {
     return new Epoll();
   }
 
-  public abstract ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor);
+  public final ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
+    return forDescriptorOnly(apiServiceDescriptor);
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: Shall we get away with this indirection?
Implementer can always overwrite this public method (if we drop the final).

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 2]
```diff
@@ -70,13 +92,42 @@ public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
...
+
+    @Override
+    protected ManagedChannel forDescriptorOnly(ApiServiceDescriptor apiServiceDescriptor) {
+      return forDescriptor(apiServiceDescriptor, interceptors);
+    }
+
+    @Override
+    protected ManagedChannel forDescriptor(
+        ApiServiceDescriptor apiServiceDescriptor, List<ClientInterceptor> interceptors) {
+      return channelFactory.forDescriptor(apiServiceDescriptor, interceptors);
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: The behavior in not well defined.
Though this is an intercepted factory, the original interceptors (interceptors from constructor) are not applied.

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 3]
```diff
@@ -70,13 +92,42 @@ public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
...
+    }
+
+    @Override
+    public ManagedChannelFactory withInterceptors(List<ClientInterceptor> interceptors) {
+      return new InterceptedManagedChannelFactory(channelFactory, interceptors);
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: I suppose this method is for Factory chaining.
Should we pass "this" in that case?

I would suggest to remove chaining in this manner and provide a builder to do the chaining.
Something like this:
`ManagedChannelFactoryBuilder.builder(channelFactory).withInterceptors(interceptors)`

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 4]
```diff
@@ -41,7 +42,24 @@ public static ManagedChannelFactory createEpoll() {
     return new Epoll();
   }
 
-  public abstract ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor);
+  public final ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
+    return forDescriptorOnly(apiServiceDescriptor);
+  }
+
+  protected ManagedChannel forDescriptorOnly(ApiServiceDescriptor descriptor) {
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: Based on the previous comment, we can remove this method.

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 5]
```diff
@@ -70,13 +92,42 @@ public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
...
           .usePlaintext(true)
           // Set the message size to max value here. The actual size is governed by the
           // buffer size in the layers above.
           .maxInboundMessageSize(Integer.MAX_VALUE)
+          .intercept(interceptors)
           .build();
     }
   }
+
+  private static class InterceptedManagedChannelFactory extends ManagedChannelFactory {
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: What is the purpose of this class?
Based on method `ManagedChannelFactory.forDescriptor(ApiServiceDescriptor apiServiceDescriptor, List<ClientInterceptor> interceptors)`
All implementations of ManagedChannelFactory should support interception and if not, they should throw exception.

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java [hunk 6]
```diff
@@ -70,13 +92,42 @@ public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
    */
   private static class Default extends ManagedChannelFactory {
     @Override
-    public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
+    public ManagedChannel forDescriptor(
```
Comment (angoenka) [inline sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: As this method requires more than just descriptor, shall we rename it to something like `createChannel` or `getChannel`?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java [hunk 7]
```diff
@@ -80,23 +83,28 @@ public static void main(String[] args) throws Exception {
     System.out.format("Control location %s%n", System.getenv(CONTROL_API_SERVICE_DESCRIPTOR));
     System.out.format("Pipeline options %s%n", System.getenv(PIPELINE_OPTIONS));
 
-    ObjectMapper objectMapper = new ObjectMapper().registerModules(
-        ObjectMapper.findModules(ReflectHelpers.findClassLoader()));
-    PipelineOptions options = objectMapper.readValue(
-        System.getenv(PIPELINE_OPTIONS), PipelineOptions.class);
+    String id = System.getenv(HARNESS_ID);
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: Shall we also log this parameter as other parameters.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java [hunk 8]
```diff
@@ -115,43 +124,46 @@ public static void main(PipelineOptions options,
...
-        options,
-        loggingApiServiceDescriptor,
-        channelFactory::forDescriptor)) {
+    try (BeamFnLoggingClient logging =
+        new BeamFnLoggingClient(
+            options, loggingApiServiceDescriptor, channelFactory::forDescriptor)) {
 
       LOG.info("Fn Harness started");
-      EnumMap<BeamFnApi.InstructionRequest.RequestCase,
+      EnumMap<
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: Please reformat the file.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java [hunk 9]
```diff
@@ -115,43 +124,46 @@ public static void main(PipelineOptions options,
...
       Endpoints.ApiServiceDescriptor loggingApiServiceDescriptor,
       Endpoints.ApiServiceDescriptor controlApiServiceDescriptor,
       ManagedChannelFactory channelFactory,
       StreamObserverFactory streamObserverFactory) {
     IdGenerator idGenerator = IdGenerators.decrementingLongs();
-    try (BeamFnLoggingClient logging = new BeamFnLoggingClient(
-        options,
-        loggingApiServiceDescriptor,
-        channelFactory::forDescriptor)) {
+    try (BeamFnLoggingClient logging =
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: logging is not used. Shall we remove it?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java [hunk 10]
```diff
@@ -6,31 +6,33 @@
...
+import io.grpc.stub.MetadataUtils;
+
+/** A {@link ClientInterceptor} that attaches a provided SDK Harness ID to outgoing messages. */
+public class AddHarnessIdInterceptor {
+  private static final Key<String> ID_KEY = Key.of("worker_id", Metadata.ASCII_STRING_MARSHALLER);
 
-  @Override
-  public ManagedChannel forDescriptor(ApiServiceDescriptor apiServiceDescriptor) {
-    return InProcessChannelBuilder.forName(apiServiceDescriptor.getUrl()).build();
+  public static ClientInterceptor create(String harnessId) {
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java]: Shall we have a null check for harnessId?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java [hunk 11]
```diff
@@ -6,31 +6,33 @@
...
- * <p>The channel builder uses {@link ApiServiceDescriptor#getUrl()} as the unique in-process name.
- */
-public class InProcessManagedChannelFactory extends ManagedChannelFactory {
+import io.grpc.ClientInterceptor;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.stub.MetadataUtils;
+
+/** A {@link ClientInterceptor} that attaches a provided SDK Harness ID to outgoing messages. */
+public class AddHarnessIdInterceptor {
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java]: Not sure if we need a repository of interceptors. Just putting it as a thought.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java [hunk 12]
```diff
@@ -69,21 +72,23 @@
...
-      Endpoints.ApiServiceDescriptor apiServiceDescriptor,
-      Function<Endpoints.ApiServiceDescriptor, ManagedChannel> channelFactory,
+      String id,
+      ApiServiceDescriptor apiServiceDescriptor,
+      ManagedChannelFactory channelFactory,
       BiFunction<
-              StreamObserverClientFactory<InstructionRequest, BeamFnApi.InstructionResponse>,
-              StreamObserver<BeamFnApi.InstructionRequest>,
-              StreamObserver<BeamFnApi.InstructionResponse>>
+              StreamObserverClientFactory<InstructionRequest, InstructionResponse>,
```
Comment (angoenka) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java]: Reformat?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java (hunk 7 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: Done.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java (hunk 8 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: Done.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java (hunk 9 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: No. This is done to close the client automatically at the completion of the `try` block, and the client must exist to send intercepted LOG messages to the logging server.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java (hunk 10 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java]: Done.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java (hunk 11 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java]: I find this much more readable as a factory method than as an inline key and `MetadataUtils` call - it tells us exactly the thing it does.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java (hunk 12 above)
Comment (tgroh) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/BeamFnControlClient.java]: This is autoformatted.

File: sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java (hunk 3 above)
Comment (youngoli) [reply sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/channel/ManagedChannelFactory.java]: Yes, I'm also curious about this. If it's not for factory chaining and isn't necessary for another reason, could it be removed?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java (hunk 9 above)
Comment (angoenka) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnHarness.java]: Can we add a comment explaining this as logging is unused which makes this code block non intuitive.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java (hunk 11 above)
Comment (angoenka) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/AddHarnessIdInterceptor.java]: SG

**Comment by jkff** [comment]
This PR is replaced by #5680.

**Pull request closed without merging**